### PR TITLE
Add IVS APIs for package indexing (part 1)

### PR DESCRIPTION
**Pull request by joelverhagen**
https://github.com/NuGet/Home/issues/2148

Adds:
- Parse framework name
- Get nearest framework
- Compare semantic versions

These are the low hanging fruit.

@emgarten @zhili1208 @deepakaravindr @yishaigalatzer

### CODE REVIEW

**Comment by emgarten** [comment]
LGTM :rocket:

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsFrameworkCompatibility.cs [hunk 1]
```diff
@@ -43,17 +44,33 @@ public IEnumerable<FrameworkName> GetFrameworksSupportingNetStandard(FrameworkNa
...
-        }
+            if (targetFramework == null)
+            {
+                throw new ArgumentNullException(nameof(targetFramework));
+            }
 
-        private FrameworkName GetFrameworkName(NuGetFramework nuGetFramework)
-        {
-            return new FrameworkName(nuGetFramework.DotNetFrameworkName);
+            if (frameworks == null)
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsFrameworkCompatibility.cs]: test for empty enumerable as well, and throw.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsFrameworkParser.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        public FrameworkName ParseFrameworkName(string shortOrFullName)
+        {
+            if (shortOrFullName == null)
+            {
+                throw new ArgumentNullException(nameof(shortOrFullName));
+            }
+
+            try
+            {
+                var nugetFramework = NuGetFramework.Parse(shortOrFullName);
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsFrameworkParser.cs]: thought - make NuGetFramework throw the right exception, and avoid the catch and cleanup here. Document on it that this is a public API

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsSemanticVersionComparer.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            }
+
+            NuGetVersion parsedVersionA;
+            if (!NuGetVersion.TryParse(versionA, out parsedVersionA))
+            {
+                string message = string.Format(
+                    CultureInfo.CurrentCulture,
+                    VsResources.InvalidSemanticVersionStringIncludingInput,
+                    versionA);
+                throw new ArgumentException(message, nameof(versionA));
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsSemanticVersionComparer.cs]: see comment above

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Utility/FrameworkNameUtility.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Runtime.Versioning;
+using NuGet.Frameworks;
+
+namespace NuGet.VisualStudio.Implementation.Utility
+{
+    public static class FrameworkNameUtility
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Utility/FrameworkNameUtility.cs]: this whole class seems redundant

File: src/NuGet.Clients/VisualStudio/Extensibility/IVsSemanticVersionComparer.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        /// <summary>
+        /// Compares two version strings as if they were NuGet semantic version
+        /// strings. Returns a number less than zero if <see cref="versionA"/>
+        /// is less than <see cref="versionB"/>. Returns zero if the two versions 
+        /// are equivalent. Returns a number greater than zero if <see cref="versionA"/>
+        /// is greater than <see cref="versionB"/>.
+        /// </summary>
+        /// <param name="versionA">The first version string.</param>
+        /// <param name="versionB">The second version string.</param>
+        /// <exception cref="ArgumentNullException">If either version string is null.</exception>
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/VisualStudio/Extensibility/IVsSemanticVersionComparer.cs]: add unit tests to validate the exception types on the public API surface.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Utility/FrameworkNameUtility.cs (hunk 4 above)
Comment (joelverhagen) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Utility/FrameworkNameUtility.cs]: Alternatives? This is a sort of mapping that is now needed on two different of our IVS interfaces. Best to have the mapping in one place, right?

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsFrameworkCompatibility.cs (hunk 1 above)
Comment (joelverhagen) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsFrameworkCompatibility.cs]: The inner implementation accepts an empty list, always returning null. So it seems like a contrived restriction at this level. Also, this is something the caller can very easily throw based on (if he so desires). Could a reasonable compromise be to explicitly state in the XML docs that an empty list is accepted?

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Utility/FrameworkNameUtility.cs (hunk 4 above)
Comment (joelverhagen) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Utility/FrameworkNameUtility.cs]: Spoke offline -- will move this mapping to the NuGetFramework type itself.

Comment (joelverhagen) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Utility/FrameworkNameUtility.cs]: After inspecting the code, the existing pattern is to use `new FrameworkName(NuGetFramework.DotNetFrameworkName)` and `NuGetFramework.Parse(FrameworkName.ToString())`. These are simple solutions so I will remain consistent and do it this way (and remove my utility thing).

File: src/NuGet.Clients/VisualStudio/Extensibility/IVsSemanticVersionComparer.cs (hunk 5 above)
Comment (joelverhagen) [inline src/NuGet.Clients/VisualStudio/Extensibility/IVsSemanticVersionComparer.cs]: Done.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsSemanticVersionComparer.cs (hunk 3 above)
Comment (joelverhagen) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsSemanticVersionComparer.cs]: Turns out `NuGetVersion.Parse` already throws `ArgumentException`. Fixing.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsFrameworkParser.cs (hunk 2 above)
Comment (joelverhagen) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsFrameworkParser.cs]: Fixed.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Utility/FrameworkNameUtility.cs (hunk 4 above)
Comment (joelverhagen) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Utility/FrameworkNameUtility.cs]: Fixed.

**Pull request closed without merging**